### PR TITLE
Add irJS Distribution console tool

### PR DESCRIPTION
**Pull request by dagardiner**
Add tool that allows irDistribution to be called from a command line - IE for use in a CI/CD pipeline

### CODE REVIEW

File: Developer Samples/BuildIrJsRuleApp/BuildIrJsRuleApp/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("BuildIrJsRuleApp")]
+[assembly: AssemblyDescription("Command line utility for requesting compiled irJS Rule Applications from the irDistribution compilation service")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("InRule Technologies")]
```
Comment (rrinaldi) [inline Developer Samples/BuildIrJsRuleApp/BuildIrJsRuleApp/Properties/AssemblyInfo.cs]: Our legal name is InRule Technology, not InRule Technologies.

File: Developer Samples/BuildIrJsRuleApp/BuildIrJsRuleApp/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+[assembly: AssemblyProduct("BuildIrJsRuleApp")]
+[assembly: AssemblyCopyright("Copyright ©  2019 InRule Technologies")]
```
Comment (rrinaldi) [inline Developer Samples/BuildIrJsRuleApp/BuildIrJsRuleApp/Properties/AssemblyInfo.cs]: Our legal name is InRule Technology, not InRule Technologies.

File: Developer Samples/BuildIrJsRuleApp/BuildIrJsRuleApp/Properties/AssemblyInfo.cs (hunk 1 above)
Comment (rrinaldi) [inline Developer Samples/BuildIrJsRuleApp/BuildIrJsRuleApp/Properties/AssemblyInfo.cs]: Our legal name is InRule Technology, not InRule Technologies.

File: Developer Samples/BuildIrJsRuleApp/BuildIrJsRuleApp/Properties/AssemblyInfo.cs (hunk 2 above)
Comment (rrinaldi) [inline Developer Samples/BuildIrJsRuleApp/BuildIrJsRuleApp/Properties/AssemblyInfo.cs]: Our legal name is InRule Technology, not InRule Technologies.

File: README.md [hunk 3]
```diff
@@ -19,6 +19,7 @@ If you would like to request additional samples from InRule, please write an iss
 | Add an existing entity to a collection | Developer Samples | AddCollectionMembers <br> CreateEntityFromXml |
 | Authoring Tutorial | Authoring Samples | irAuthor Beginner Tutorial |
 | Auto Rule Sets | Authoring Samples | Iteration |
+| Build irJS Compiled Rule App from itDistribution Service using Command Line | Developer Samples | BuildIrJsRuleApp |
```
Comment (rrinaldi) [inline README.md]: Typo in irDistribution

File: Developer Samples/BuildIrJsRuleApp/BuildIrJsRuleApp/Properties/AssemblyInfo.cs (hunk 1 above)
Comment (dagardiner) [reply Developer Samples/BuildIrJsRuleApp/BuildIrJsRuleApp/Properties/AssemblyInfo.cs]: File (and concept of assembly info) was removed during conversion to .net core

Comment (dagardiner) [reply Developer Samples/BuildIrJsRuleApp/BuildIrJsRuleApp/Properties/AssemblyInfo.cs]: File (and concept of assembly info) was removed during conversion to .net core

File: README.md (hunk 3 above)
Comment (dagardiner) [reply README.md]: File (and concept of assembly info) was removed during conversion to .net core

[the remark this answers is not in the document]
Comment (dagardiner) [reply README.md]: Good catch, updated

File: Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                {
+                    client.BaseAddress = new Uri(serviceUri);
+
+                    // Build up our request by reading in the rule application
+                    var ruleApplication = ruleApplicationRef.GetRuleApplicationDef();
+                    var httpContent = new ByteArrayContent(Encoding.UTF8.GetBytes(ruleApplication.GetXml()));
+                    requestContent.Add(httpContent, "ruleApplication", ruleApplication.Name + ".ruleapp");
+
+                    // Tell the server we are sending form data
+                    client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("multipart/form-data"));
```
Comment (jelster) [inline Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs]: Might want to also add the appropriate HTTP header (I forget the exact one OTOMH) for telling the server to keep the connection alive while waiting for the service to respond

File: Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+                    client.DefaultRequestHeaders.Add("subscription-key", subscriptionKey);
+
+                    // Post the rule application to the irDistribution service API,
+                    // enabling Execution and State Change logging, Display Name metadata, and the developer example.
+                    var distributionUrl = "package?logOption=Execution&logOption=StateChanges&subscriptionkey=" + subscriptionKey;
+                    result = await client.PostAsync(distributionUrl, requestContent).ConfigureAwait(false);
+
+                    // Get the return package from the result
+
+                    dynamic returnPackage = JObject.Parse(result.Content.ReadAsStringAsync().Result);
```
Comment (jelster) [inline Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs]: no reason not to `await` this; declare and init a separate variable first.

File: Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+                    requestContent.Add(httpContent, "ruleApplication", ruleApplication.Name + ".ruleapp");
+
+                    // Tell the server we are sending form data
+                    client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("multipart/form-data"));
+                    client.DefaultRequestHeaders.Add("subscription-key", subscriptionKey);
+
+                    // Post the rule application to the irDistribution service API,
+                    // enabling Execution and State Change logging, Display Name metadata, and the developer example.
+                    var distributionUrl = "package?logOption=Execution&logOption=StateChanges&subscriptionkey=" + subscriptionKey;
+                    result = await client.PostAsync(distributionUrl, requestContent).ConfigureAwait(false);
```
Comment (jelster) [inline Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs]: you're already using `await` here, why are you calling `.ConfigureAwait(false)`?

File: Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+                        }
+                        // Still need to stop processing
+                        return errors.ToString();
+                    }
+
+                    // Build the download url of the file
+                    var downloadUrl = returnPackage.PackagedApplicationDownloadUrl.ToString();
+
+                    // Get the contents
+                    HttpResponseMessage resultDownload = await client.GetAsync(downloadUrl).ConfigureAwait(false);
```
Comment (jelster) [inline Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs]: same question about `.ConfigureAwait`

File: Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+                    if (!resultDownload.IsSuccessStatusCode)
+                    {
+                        // Handle errors
+                        errors.AppendLine(resultDownload.Content.ReadAsStringAsync().Result);
```
Comment (jelster) [inline Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs]: similar to above, using await is recommended

File: Developer Samples/CommandLineTools/BuildIrJsRuleApp/BuildIrJsRuleApp.csproj [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    <TargetFramework>netcoreapp2.2</TargetFramework>
+    <AssemblyName>BuildIrJsRuleApp</AssemblyName>
+    <ApplicationIcon />
+    <StartupObject />
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="InRule.Repository" Version="5.4.0" />
+    <PackageReference Include="InRule.Runtime" Version="5.4.0" />
+    <PackageReference Include="Mono.Options" Version="5.3.0.1" />
```
[the remark this answers is not in the document]
Comment (dagardiner) [reply Developer Samples/CommandLineTools/BuildIrJsRuleApp/BuildIrJsRuleApp.csproj]: The functionality is there, but it's not as clean as Mono, so I used that instead.  Ref - https://msdn.microsoft.com/en-us/magazine/mt833289.aspx

File: Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+            }
+            catch (OptionException e)
+            {
+                Console.Write("Failed parsing execution parameters: " + e.Message);
+                showHelp = true;
+            }
+
+            if (showHelp)
+            {
+                ShowHelp(clParams);
```
[the remark this answers is not in the document]
Comment (dagardiner) [reply Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs]: Stylistically, I dislike having multiple exit points from a method for something this simple - this structure keeps it more linear and easier to follow.

File: Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+            }
+            else if (string.IsNullOrEmpty(irDistributionKey))
+            {
+                Console.WriteLine("Error: Missing required parameter DistributionKey.");
```
[the remark this answers is not in the document]
Comment (dagardiner) [reply Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs]: see above

File: Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+            }
+            else if (string.IsNullOrEmpty(OutputFilePath))
+            {
+                Console.WriteLine("Error: Missing required parameter OutputPath.");
+            }
+            else if (!string.IsNullOrEmpty(RuleAppFilePath))
```
[the remark this answers is not in the document]
Comment (dagardiner) [reply Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs]: updated

File: Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+            Console.WriteLine();
+            Console.WriteLine("All requests must contain DistributionKey and OutputPath.");
+            Console.WriteLine("   File-based Rule Apps should be specified with RuleAppPath.");
+            Console.WriteLine("   Catalog-based Rule Apps should be specified with CatalogUri, CatalogUsername, CatalogPassword, and CatalogRuleAppName (optionally CatalogRuleAppLabel).");
+            Console.WriteLine();
+            Console.WriteLine("Options:");
+            p.WriteOptionDescriptions(Console.Out);
+            Console.WriteLine();
+        }
+        private static void RetrieveAndWriteIrJSFromDistributionService(RuleApplicationReference ruleRef, string distroKey, string outputPath)
```
[the remark this answers is not in the document]
Comment (dagardiner) [reply Developer Samples/CommandLineTools/BuildIrJsRuleApp/Program.cs]: updated

**Pull request merged**